packages/legend-query/src/components/QueryBuilderValueSpecificationEditor.tsx

```diff
@@ -533,75 +587,596 @@ export const LatestDatePrimitiveInstanceValueEditor: React.FC = () => (
+                graph,
+              ),
+            )
+          : updateValueSpecification(
+              generateDateFunctionExpression(graph, dateOptionValue),
+            );
       }
+      setDateOptionValue({ ...dateOptionValue });
     };
 
     return (
       <div className="query-builder-value-spec-editor__date">
-        {valueSpecification.genericType.value.rawType === strictDate && (
+        {typeCheckOption.expectedType === strictDate && (
           <DatePrimitiveInstanceValueEditor
-            valueSpecification={valueSpecification}
+            valueSpecification={valueSpecification as PrimitiveInstanceValue}
             className={className}
           />
         )}
-        {valueSpecification.genericType.value.rawType === dateTime && (
-          <DateTimePrimitiveInstanceValueEditor
-            valueSpecification={valueSpecification}
-            className={className}
-          />
-        )}
-        {valueSpecification.genericType.value.rawType === latestDate && (
+        {typeCheckOption.expectedType === latestDate && (
           <LatestDatePrimitiveInstanceValueEditor />
         )}
-        {expectedType === date && (
-          <div className="query-builder-value-spec-editor__dropdown">
-            <CustomSelectorInput
-              placeholder="Choose a type..."
-              options={typeOptions}
-              onChange={changeType}
-              value={selectedType}
-              darkMode={true}
-            />
+        {(typeCheckOption.expectedType === date ||
+          typeCheckOption.expectedType === dateTime) && (
+          <div className="query-builder-value-spec-editor__date-option-dropdown">
+            <button
+              className="query-builder-value-spec-editor__date-option-dropdown__button"
+              onClick={openDateOptionsPopover}
+            >
+              {/* for the reset icon to work*/}
+              {generateDateOption(valueSpecification, graph).label}
+            </button>
+            <QueryBuilderDateOptionsInfoTooltip>
+              <div className="query-builder-value-spec-editor__date-option-dropdown__info-icon">
+                <InfoCircleIcon />
+              </div>
+            </QueryBuilderDateOptionsInfoTooltip>
+            <BasePopover
+              open={Boolean(anchorEl)}
+              anchorEl={anchorEl}
+              onClose={handleCloseDateOptionsPopover}
+              anchorOrigin={{
+                vertical: 'bottom',
+                horizontal: 'center',
+              }}
+              transformOrigin={{
+                vertical: 'top',
+                horizontal: 'center',
+              }}
+            >
+              <BaseRadioGroup
+                value={dateOptionValue.value}
+                onChange={handleDateOptionChange}
+                row={true}
+                options={targetQueryBuilderDateOptionsEnum}
+                size={2}
+              />
+              {dateOptionValue.value ===
```

Comment:
Could we leverage case statements for this rendering. Becomes very unclear when we have multiply && statements like this as to what is being rendered in the screen